src/main/java/com/vmware/vipclient/i18n/messages/service/ComponentsService.java

```diff
@@ -129,4 +153,38 @@ private Map<String, String> makeLocaleMap(final SortedSet<String> localesToQuery
 
         return map;
     }
+
+    private void getFallbackLocaleMessages(Iterator<Locale> fallbackLocalesIter, final Set<String> componentsToFallback,
+                                                                       final Set<String> localesToFallback, final Map<String, Map<String, Map<String, String>>> dataMap) {
+        if (fallbackLocalesIter != null && fallbackLocalesIter.hasNext()) {
+            Locale fallbackLocale = fallbackLocalesIter.next();
+            if (fallbackLocale.toLanguageTag().equals(ConstantsKeys.SOURCE)) {
+                fallbackLocale = Locale.forLanguageTag(ConstantsKeys.LATEST);
+            }
+            Set<Locale> localeSet = new HashSet<Locale>();
+            localeSet.add(fallbackLocale);
+            Map<Locale, Map<String, Map<String, String>>> localeDataMap = getTranslation(componentsToFallback, localeSet, fallbackLocalesIter);
+            Map<String, Map<String, String>> componentsMap = localeDataMap.get(fallbackLocale);
+            if (componentsMap != null && !componentsMap.isEmpty()) {
+                final SortedSet<String> localesToFallbackAgain = new TreeSet<>();
+                final SortedSet<String> componentsToFallbackAgain = new TreeSet<>();
+                for (String locale : localesToFallback) {
+                    for (String component : componentsToFallback) {
+                        Map<String, String> componentMap = componentsMap.get(component);
+                        if (componentMap != null) {
+                            if (dataMap.get(locale).get(component) == null)
+                                dataMap.get(locale).put(component, componentMap);
```

Comment:
```suggestion
                for (String component : componentsToFallback) {
                    Map<String, String> componentMap = componentsMap.get(component);
                    for (String locale : localesToFallback) {
                        if (componentMap != null && (dataMap.get(locale).get(component) == null)) {
                            dataMap.get(locale).put(component, componentMap);```

Reply:
No, if change as what you suggest, the logic will be incorrect. The conditions of judging fallback failure or not are:
dataMap.get(locale).get(component) == null **and** componentMap == null instead of:
dataMap.get(locale).get(component) != null **or** componentMap == null

Also optimize code to make it more readable.